sway/input/text_input.c

```diff
@@ -26,6 +32,329 @@ static struct sway_text_input *relay_get_focused_text_input(
+	}
+
+	int popup_width = popup->popup_surface->surface->current.width;
+	int popup_height = popup->popup_surface->surface->current.height;
+	int x1 = parent.x + cursor.x;
+	int x2 = parent.x + cursor.x + cursor.width;
+	int y1 = parent.y + cursor.y;
+	int y2 = parent.y + cursor.y + cursor.height;
+	int x = x1;
+	int y = y2;
+
+	int available_right = output_box.x + output_box.width - x1;
+	int available_left = x2 - output_box.x;
+	if (available_right < popup_width && available_left > available_right) {
+		x = x2 - popup_width;
+	}
+
+	int available_down = output_box.y + output_box.height - y2;
+	int available_up = y1 - output_box.y;
+	if (available_down < popup_height && available_up > available_down) {
+		y = y1 - popup_height;
+	}
+
+	popup->x = x - parent.x;
+	popup->y = y - parent.y;
+
+	// Hide popup if cursor position is completely out of bounds
+	bool x1_in_bounds = (cursor.x >= 0 && cursor.x < parent.width);
+	bool y1_in_bounds = (cursor.y >= 0 && cursor.y < parent.height);
+	bool x2_in_bounds = (cursor.x + cursor.width >= 0
+		&& cursor.x + cursor.width < parent.width);
+	bool y2_in_bounds = (cursor.y + cursor.height >= 0
+		&& cursor.y + cursor.height < parent.height);
+	popup->visible =
+		(x1_in_bounds && y1_in_bounds) || (x2_in_bounds && y2_in_bounds);
+
+	if (cursor_rect) {
+		struct wlr_box box = {
+			.x = x1 - x,
+			.y = y1 - y,
+			.width = cursor.width,
+			.height = cursor.height,
+		};
+		wlr_input_popup_surface_v2_send_text_input_rectangle(
+			popup->popup_surface, &box);
+	}
+
+	sway_input_popup_damage(popup);
+}
+
+static void surface_send_enter_iterator(struct wlr_surface *surface,
+		int x, int y, void *data) {
+	struct wlr_output *wlr_output = data;
+	wlr_surface_send_enter(surface, wlr_output);
+}
+
+static void surface_send_leave_iterator(struct wlr_surface *surface,
+		int x, int y, void *data) {
+	struct wlr_output *wlr_output = data;
+	wlr_surface_send_leave(surface, wlr_output);
+}
+
+static void input_popup_send_outputs(struct sway_input_popup *popup,
+		wlr_surface_iterator_func_t iterator) {
+	struct sway_text_input *text_input =
+		relay_get_focused_text_input(popup->relay);
+	if (text_input == NULL || text_input->input->focused_surface == NULL) {
+		return;
+	}
+	struct wlr_surface *focused_surface = text_input->input->focused_surface;
+	struct wlr_layer_surface_v1 *layer_surface =
+		wlr_layer_surface_v1_try_from_wlr_surface(focused_surface);
+	if (layer_surface != NULL) {
+		struct sway_layer_surface *layer =
+			layer_from_wlr_layer_surface_v1(layer_surface);
+		wlr_surface_for_each_surface(popup->popup_surface->surface,
+			iterator, layer->layer_surface->output);
+		return;
+	}
+	struct sway_view *view = view_from_wlr_surface(focused_surface);
+	for (int i = 0; i < view->container->outputs->length; i++) {
+		struct sway_output *output = view->container->outputs->items[i];
+		wlr_surface_for_each_surface(popup->popup_surface->surface,
+			iterator, output->wlr_output);
+	}
```

Comment:
Why is only one output send in the case of a `layer_surface` but all of them are sent in the other case?

Reply:
sway_view cannot get the output the surface in, and there is the case that the surface is both in two outputs, so the author before think handle like this is ok. but layer_shell will just in one wl_output